STYLE: turn off PLW2901 (but keep some changes)

ErdiTk opened the pull request:
Continuation of work started on closed issue #51708.
Didn't find a new issue to link this PR, may I ask you @MarcoGorelli if it is fine to link it to the closed one or shall I wait for a new issue to be opened?

ErdiTk committed:
Removed unnecessary code

ErdiTk committed:
Enabled PLW2901 on initial files under pandas/core/*

MarcoGorelli reviewed (changes requested):
thanks

can you update the pyproject.toml file accordingly please?

ErdiTk commented:
> thanks
> 
> can you update the pyproject.toml file accordingly please?

I will update it when I finish all the files under  `pandas/core/*`. Give me few days for it, there are quite some to change :)

MarcoGorelli commented:
it's fine (and preferable) split it up into smaller chunks

we can split up the ` pandas/core/*` ignore into multiple ones

ErdiTk committed:
Divided pandas/core/* in pyproject.toml. Included some more fixes.

ErdiTk committed:
Merge branch 'main' into gradually_enable_plw2901

MarcoGorelli reviewed (changes requested):
looks like the "
Code Checks / Docstring validation, typing, and other manual pre-commit hooks (pull_request) " is failing - code you give me a ping when that's one's fixed please?

ErdiTk commented:
> looks like the "
> Code Checks / Docstring validation, typing, and other manual pre-commit hooks (pull_request) " is failing - code you give me a ping when that's one's fixed please?

Sure, I think I know which changes might have impacted it. I will ping you around Tuesday, unfortunately I won't make it earlier.

MarcoGorelli commented:
sure, no hurry, thanks!

ErdiTk committed:
Not sure why but the call to partial removes the .__name__ attribute to aggfunc if just returned.

ErdiTk committed:
Merge branch 'main' into gradually_enable_plw2901

ErdiTk commented on pandas/core/apply.py:
```diff
@@ -1427,14 +1427,17 @@ def _managle_lambda_list(aggfuncs: Sequence[Any]) -> Sequence[Any]:
+            mangled_aggfuncs.append(aggfunc)
+            continue
+
+        partial_aggfunc = aggfunc
```
@MarcoGorelli by adding this line the error should be solved. At least locally the mypy error is gone. Not sure why but by simply creating `partial_aggfunc` from the call to `partial(aggfunc)` in the next line it removes the `__name__` attribute. But if I firstly assign it as equal to `aggfunc` as I am doing here it works fine. Do you know any similar behavior?

ErdiTk replied:
Oh thank you a lot for the suggestion! I will implement it like this. Sorry but I am still learning how to properly handle those issues.

MarcoGorelli commented on pandas/core/apply.py:
```diff
@@ -1427,14 +1427,17 @@ def _managle_lambda_list(aggfuncs: Sequence[Any]) -> Sequence[Any]:
+            continue
+
+        partial_aggfunc = aggfunc
+        partial_aggfunc = partial(aggfunc)
+        partial_aggfunc.__name__ = f"<lambda_{idx}>"
```
also, I don't think the logic is right here, please try to keep the logic the same as the original

ErdiTk replied:
The original one was:
```
i = 0
for aggfunc in aggfuncs:
        if com.get_callable_name(aggfunc) == "<lambda>":
            aggfunc = partial(aggfunc)
            aggfunc.__name__ = f"<lambda_{i}>"
            i += 1
        mangled_aggfuncs.append(aggfunc)
```

I removed the variable `i`, replacing it by `idx` generated by the call to `enumerate`. Then I inverted the `if` clause, making it a `if not`. In this way I can insert inside the condition the single line:
`mangled_aggfuncs.append(aggfunc)`
and then continue the loop. So I moved what was inside the if outside. Sorry I understand it's not super clear, but I thought would look nicer to have a lighter content inside the if statement. If you want I can revert the logic.

MarcoGorelli reviewed (changes requested):
thanks for your PR

I've left a couple of comments, there should be enough to work on here

ErdiTk committed:
Ignored type checker and corrected possible induced bug in period.py

ErdiTk committed:
Merge branch 'main' into gradually_enable_plw2901

ErdiTk commented on pandas/core/arrays/period.py:
```diff
@@ -1105,7 +1105,7 @@ def _range_from_fields(
         for y, q in zip(year, quarter):
-            y, m = parsing.quarter_to_myear(y, q, freqstr)
+            _, m = parsing.quarter_to_myear(y, q, freqstr)
             val = libperiod.period_ordinal(y, m, 1, 1, 1, 1, 0, 0, base)
```
My bad, indeed the variable `y` can be changed inside the function `quarter_to_myear`. Updated! Thanks for the heads up!

ErdiTk committed:
Forgot to add the not in previous commit

MarcoGorelli commented on pandas/core/apply.py:
```diff
@@ -1427,14 +1427,17 @@ def _managle_lambda_list(aggfuncs: Sequence[Any]) -> Sequence[Any]:
-            aggfunc.__name__ = f"<lambda_{i}>"
-            i += 1
-        mangled_aggfuncs.append(aggfunc)
+    for idx, aggfunc in enumerate(aggfuncs):
+        if not com.get_callable_name(aggfunc) == "<lambda>":
+            mangled_aggfuncs.append(aggfunc)
+            continue
+
+        partial_aggfunc = partial(aggfunc)
+        # "partial[Any]" has no attribute "__name__"
+        partial_aggfunc.__name__ = f"<lambda_{idx}>"  # type: ignore[attr-defined]
+        mangled_aggfuncs.append(partial_aggfunc)
```
this isn't the same logic - before, `i` was only incremented if `com.get_callable_name(aggfunc) == "<lambda>"`, whereas now you're incrementing `idx` on every iteration

MarcoGorelli commented on pandas/core/arrays/period.py:
```diff
@@ -1105,8 +1105,8 @@ def _range_from_fields(
         for y, q in zip(year, quarter):
-            y, m = parsing.quarter_to_myear(y, q, freqstr)
-            val = libperiod.period_ordinal(y, m, 1, 1, 1, 1, 0, 0, base)
+            y_parsed, m = parsing.quarter_to_myear(y, q, freqstr)
```
maybe `calendar_year, calendar_month = parsing.quarter_to_myear(y, q, freqstr)`?

MarcoGorelli reviewed (changes requested):
thanks for working on this

I'm going to go back to my previous comment https://github.com/pandas-dev/pandas/pull/51797#pullrequestreview-1327035638 , I'm not sure we want to enable this in CI

some of these changes are nice

shall we keep the
```
where[idx] = _validate_where(w)
```
change (which is quite nice), revert the rest, and move `PLW2901` to the 'intentionally turned off' section of pyproject.toml?

https://github.com/pandas-dev/pandas/blob/af55880968d43c907dcf9b679f1ce227d76c143e/pyproject.toml#L316

ErdiTk committed:
Reverted changes and updated pyproject.toml

ErdiTk committed:
Merge branch 'main' into gradually_enable_plw2901

ErdiTk commented:
Ok, I reverted most of the changes except on few files.
Under `# intentionally turned off ` I added `"redefined-loop-name"`, but I saw the errors still appearing when running `pre-commit run ruff --all-files`, so I kept `"PLW2901"` also under the to ignore section of ruff. Please let me know if this is the correct way. Also let me know if the few changes listed are fine.

Commit:
turn off redefined-loop-variable, minimise diff

Commit:
revert fun_name change

MarcoGorelli commented on pandas/core/apply.py:
```diff
@@ -1423,10 +1423,10 @@ def relabel_result(
         # mean  1.5
         # mean  1.5
         if reorder_mask:
-            fun = [
```
not sure we can do this, `fun` is still used below

MarcoGorelli reviewed (approved):
looks fine, just made some minor fixup, thanks @ErdiTk
